config/migrate/migrate.go

```diff
@@ -49,6 +61,92 @@ func ZeroToOne(cfg *config.Config) error {
 	}
 
 	cfg.Revision = 1
+
+	if err := safeWriteConfig(cfg); err != nil {
+		rollbackConfigWrite(cfg)
+		return err
+	}
+
+	return nil
+}
+
+// OneToTwo migrates a configuration from Revision 1 to Revision 2
+func OneToTwo(cfg *config.Config) error {
+
+	// TODO(ramfox): qfs migration
+
+	if err := oneToTwoConfig(cfg); err != nil {
+		return err
+	}
+	if err := safeWriteConfig(cfg); err != nil {
+		rollbackConfigWrite(cfg)
+		return err
+	}
+	
+	return nil
+}
+
+func oneToTwoConfig(cfg *config.Config) error {
+	if cfg.API != nil {
+		apiCfg := cfg.API
+		defaultApiCfg := config.DefaultAPI()
+		if apiCfg.Address == "" {
+			apiCfg.Address = defaultApiCfg.Address
+		}
+		if apiCfg.WebsocketAddress == "" {
+			apiCfg.WebsocketAddress = defaultApiCfg.WebsocketAddress
+		}
+	} else {
+		return qerr.New(fmt.Errorf("invalid config"), "config does not contain API configuration")
+	}
+
+	if cfg.RPC != nil {
+		defaultRpcCfg := config.DefaultRPC()
+		if cfg.RPC.Address == "" {
+			cfg.RPC.Address = defaultRpcCfg.Address
+		}
+	} else {
+		return qerr.New(fmt.Errorf("invalid config"), "config does not contain RPC configuration")
+	}
+
+	cfg.Revision = 2
```

Comment:
The ZeroToOne migration has this nice pattern that looks like:

```
  // conversion stuff happens
  cfg.Revision = 1
  if err := safeWriteConfig(...)
  return nil
```

Which makes it read really nicely; the Revision being incremented is very obvious. Could we make OneToTwo resemble the same shape? So the top-level would be:

```
  if err := oneToTwoConfig(...)
  cfg.Revision = 2
  if err := cfg.Validate(...)
  if err := safeWriteConfig(...)
  return nil
```

All it would take is moving the two blocks up to the calling function

Reply:
Really like this